bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java

```diff
@@ -0,0 +1,397 @@
+        if (!SUPPORTED_THING_TYPES_UIDS.containsKey(block.moduleID)) {
+            logger.debug("ModuleID {} is not supported, skipping this block", block.moduleID);
+            return;
+        }
+
+        CommonModelBlock commonBlock = lastCommonBlock;
+
+        if (commonBlock == null) {
+            logger.warn(
+                    "Found model block without a preceding common block. Can't add device because details are unkown");
+            return;
+        }
+
+        ThingUID thingUID = new ThingUID(SUPPORTED_THING_TYPES_UIDS.get(block.moduleID), handler.getUID(),
+                Integer.toString(block.address));
+
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(PROPERTY_VENDOR, commonBlock.manufacturer);
+        properties.put(PROPERTY_MODEL, commonBlock.model);
+        properties.put(PROPERTY_SERIAL_NUMBER, commonBlock.serialNumber);
+        properties.put(PROPERTY_VERSION, commonBlock.version);
+        properties.put(PROPERTY_BLOCK_ADDRESS, block.address);
+        properties.put(PROPERTY_BLOCK_LENGTH, block.length);
+        properties.put(PROPERTY_UNIQUE_ADDRESS, handler.getUID().getAsString() + ":" + block.address);
+
+        DiscoveryResult result = DiscoveryResultBuilder.create(thingUID).withProperties(properties)
+                .withRepresentationProperty(PROPERTY_UNIQUE_ADDRESS).withBridge(handler.getUID())
+                .withLabel(commonBlock.manufacturer + " " + commonBlock.model).build();
+
+        listener.thingDiscovered(result);
+    }
+
+    /**
+     * Parsing of model blocks finished
+     * Now we have to report back to the handler the common block and the block we were looking for
+     */
+    private void parsingFinished() {
+        listener.discoveryFinished();
+    }
+
+    /**
+     * Handle errors received during communication
+     */
+    private void handleError(Exception error) {
+        String msg = "";
+        String cls = "";
+
+        if (blocksFound > 1 && error instanceof ModbusSlaveErrorResponseException) {
+            int code = ((ModbusSlaveErrorResponseException) error).getExceptionCode();
+            if (code == ModbusSlaveErrorResponseException.ILLEGAL_DATA_ACCESS
+                    || code == ModbusSlaveErrorResponseException.ILLEGAL_DATA_VALUE) {
+                // It is very likely that the slave does not report an end block (0xffff) after the main blocks
+                // so we treat this situation as normal.
+                logger.debug(
+                        "Seems like slave device does not report an end block. Continouing with the dectected blocks");
+                parsingFinished();
+                return;
+            }
+        }
+
+        cls = error.getClass().getName();
+        msg = error.getMessage();
+
+        logger.warn("Error with read at address {}: {} {}", baseAddress, cls, msg);
+
+        possibleAddresses.remove(0); // Drop the current base address, and continue with the next one
```

Comment:
If you poll within detectModel() you can remove this

Reply:
thanks, removed.